runtime/src/qio/auxFilesys/hdfs/qio_plugin_hdfs.c

```diff
@@ -147,10 +147,36 @@ qioerr hdfs_preadv (void* file, const struct iovec *vector, int count, off_t off
 
   STARTING_SLOW_SYSCALL;
 
+#ifdef HDFS3
+
+  DO_RETAIN(((hdfs_fs*)fs));
+  
+  hdfs_file hfl = *to_hdfs_file(file);
+  hdfs_fs hfs = *to_hdfs_fs(fs);
+
+  hfl.file = hdfsOpenFile(hfs.hfs, hfl.pathnm, O_RDONLY, 0, 0, 0);
```

Comment:
The result of hdfsOpenFile should be stored in a local variable, not in hfl.file (which would presumably update any other use of the QIO file to use the newly openned HDFS file - not what you intented).